File: Data Structures & Algorithms/three-integer-sum/submission-5.js
class Solution {
    /**
     * @param {number[]} nums
     * @return {number[][]}
     */
    threeSum(nums) {
        nums.sort((a, b) => a - b)
        const res = {}
        for (let i = 1; i < nums.length - 1; i ++) {
            let li = i - 1;
            let ri = i + 1;
            const target = -nums[i];
            while ((li >= 0) && (ri < nums.length)) {
                const sum = nums[li] + nums[ri];
                if (sum == target) {
                    const arr = [nums[li], nums[i], nums[ri]]
                    arr.sort((a, b) => b - a)
                    res[arr.join(',')] = true;
                    break;
                }
                if (sum < target) {
                    ri++;
                } else {
                    li--;
                }
            }
        }
        return Object.keys(res).map((v) => v.split(',').map(Number))
    }
}

File: Data Structures & Algorithms/permutations-ii/submission-5.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        n = len(nums)
        res = []

        def rec(i, arr, used):
            if i == len(nums):
                res.append(arr.copy())
                return
            for j in range(n):
                if used[j]:
                    continue
                if j > 0 and nums[j] == nums[j - 1] and (not used[j - 1]):
                    continue
                used[j] = True
                arr.append(nums[j])

                rec(i + 1, arr, used)

                used[j] = False
                arr.pop()

        rec(0, [], [False] * n)
        return res


        # [1,1,2]
        # [1,]